linkis-commons/linkis-storage/src/main/java/org/apache/linkis/storage/fs/impl/S3FileSystem.java

```diff
@@ -0,0 +1,363 @@
+      s3Client.deleteObject(bucket, newDest.getPath());
+      throw new IOException(
+          "You have not permission to access path "
+              + oldDest.getPath()
+              + " or "
+              + newDest.getPath());
+    }
+  }
+
+  @Override
+  public boolean copy(String origin, String dest) throws IOException {
+    try {
+      s3Client.copyObject(bucket, origin, bucket, dest);
+      return true;
+    } catch (AmazonS3Exception e) {
+      throw new IOException("You have not permission to access path " + origin + " or " + dest);
+    }
+  }
+
+  private boolean isDir(S3ObjectSummary s3ObjectSummary, String prefix) {
+    return s3ObjectSummary.getKey().substring(prefix.length()).contains("/");
+  }
+
+  private boolean isInitFile(S3ObjectSummary s3ObjectSummary) {
+    return s3ObjectSummary.getKey().contains(INIT_FILE_NAME);
+  }
+
+  @Override
+  public String listRoot() {
+    return "/";
+  }
+
+
+  @Override
+  public boolean mkdir(FsPath dest) throws IOException {
+    String path = new File(dest.getPath(), INIT_FILE_NAME).getPath();
+    if (exists(new FsPath(path))) {
+      return false;
+    }
+    return create(path);
+  }
+  @Override
+  public boolean mkdirs(FsPath dest) throws IOException {
+    return mkdir(dest);
+  }
+
+  private FsPath fillStorageFile(FsPath fsPath, S3ObjectSummary s3ObjectSummary) {
+    fsPath.setModification_time(s3ObjectSummary.getLastModified().getTime());
+    fsPath.setOwner(s3ObjectSummary.getOwner().getDisplayName());
+    try {
+      fsPath.setIsdir(isDir(s3ObjectSummary, fsPath.getParent().getPath()));
+    } catch (Throwable e) {
+      logger.warn("Failed to fill storage file：" + fsPath.getPath(), e);
+    }
+
+    if (fsPath.isdir()) {
+      fsPath.setLength(0);
+    } else {
+      fsPath.setLength(s3ObjectSummary.getSize());
+    }
+    return fsPath;
+  }
+
+  @Override
+  public boolean canRead(FsPath dest) {
+    return true;
+  }
+
+  @Override
+  public boolean canWrite(FsPath dest) {
+    return true;
+  }
+
+  @Override
+  public long getTotalSpace(FsPath dest) {
+    return 0;
+  }
+
+  @Override
+  public long getFreeSpace(FsPath dest) {
+    return 0;
+  }
+
+  @Override
+  public long getUsableSpace(FsPath dest) {
+    return 0;
+  }
+
+  @Override
+  public boolean canExecute(FsPath dest) {
+    return true;
+  }
+
+  @Override
+  public boolean setOwner(FsPath dest, String user, String group) {
+    return false;
+  }
+
+  @Override
+  public boolean setOwner(FsPath dest, String user) {
+    return false;
+  }
+
+  @Override
+  public boolean setGroup(FsPath dest, String group) {
+    return false;
+  }
+
+  @Override
+  public boolean setPermission(FsPath dest, String permission) {
+    return false;
+  }
+
+  @Override
+  public void close() throws IOException {}
+
+  public String getLabel() {
+    return label;
+  }
+
+  public void setLabel(String label) {
+    this.label = label;
+  }
+
+  public String buildPath(String path)   {
```

Comment:
Thanks for your advice. I use this method because the subdirectories I get through s3client don't have a "/" prefix. I tested this at minIo. The method you mentioned does not meet my needs.

Reply:
> Thanks for your advice. I use this method because the subdirectories I get through s3client don't have a "/" prefix. I tested this at minIo. The method you mentioned does not meet my needs.

ok, thanks!